soroban-env-common/src/raw_val.rs

```diff
@@ -1,101 +1,163 @@
-use stellar_xdr::{ScStatic, ScStatus, ScStatusType};
+use crate::{impl_wrapper_tryfroms_and_tryfromvals, SymbolSmall};
+use stellar_xdr::{ScStatus, ScStatusType};
 
-use super::{BitSet, Env, Object, Static, Status, Symbol, TryFromVal};
+use super::{Env, Object, Status, TryFromVal};
 use core::{convert::Infallible, fmt::Debug};
 
 extern crate static_assertions as sa;
 
+// A Soroban Val is a 64-bit word with the low 8 bits assigned to a `tag`
+// indicating its type and the high 56 bits reserved for its `body`
+
 #[allow(dead_code)]
 const WORD_BITS: usize = 64;
-const TAG_BITS: usize = 3;
+pub(crate) const TAG_BITS: usize = 8;
 const TAG_MASK: u64 = (1u64 << TAG_BITS) - 1;
-sa::const_assert!(TAG_MASK == 0x7);
+sa::const_assert!(TAG_MASK == 0xff);
 
 #[allow(dead_code)]
-pub(crate) const BODY_BITS: usize = WORD_BITS - (TAG_BITS + 1);
-sa::const_assert!(BODY_BITS == 60);
+pub(crate) const BODY_BITS: usize = WORD_BITS - TAG_BITS;
+sa::const_assert!(BODY_BITS == 56);
+
+// The body is sometimes further subdivided into two fields:
+// a 32-bit `major` part and a 24-bit `minor` part.
 
-#[allow(dead_code)]
 const MAJOR_BITS: usize = 32;
-#[allow(dead_code)]
-const MINOR_BITS: usize = 28;
-#[allow(dead_code)]
+const MINOR_BITS: usize = 24;
 const MAJOR_MASK: u64 = (1u64 << MAJOR_BITS) - 1;
 const MINOR_MASK: u64 = (1u64 << MINOR_BITS) - 1;
 sa::const_assert!(MAJOR_MASK == 0xffff_ffff);
-sa::const_assert!(MINOR_MASK == 0x0fff_ffff);
+sa::const_assert!(MINOR_MASK == 0x00ff_ffff);
 sa::const_assert!(MAJOR_BITS + MINOR_BITS == BODY_BITS);
 
-/// Code values for the 3 "tag" bits in the bit-packed representation
-/// of [RawVal].
+/// Code values for the 8 `tag` bits in the bit-packed representation
+/// of [RawVal]. These don't coincide with tag numbers in the SCVal XDR
+/// but cover all those cases as well as some optimized refinements for
+/// special cases (boolean true and false, small-value forms).
 #[repr(u8)]
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
 pub enum Tag {
+    /// Tag for a [RawVal] that encodes [bool] `false`. The bool type is refined to
+    /// two single-value subtypes in order for each tag number to coincides with
+    /// the WASM encoding of a boolean.
+    False = 0,
+
+    /// Tag for a [RawVal] that encodes [bool] `true`.
+    True = 1,
+
+    /// Tag for a [RawVal] that is empty/absent (eg. void, null, nil, undefined, None)
+    Void = 2,
+
+    /// Tag for a [RawVal] that is contains an error code.
+    Status = 3,
+
     /// Tag for a [RawVal] that contains a [u32] number.
-    U32 = 0,
+    U32 = 4,
 
     /// Tag for a [RawVal] that contains an [i32] number.
-    I32 = 1,
+    I32 = 5,
+
+    /// Tag for a [RawVal] that contains a [u64] small enough to fit in 56 bits.
+    U64Small = 6,
```

Comment:
I really like that all the small integer types are disjoint when in their small form and don't share a single type. 👏🏻

Reply:
Yeah, I'm curious to see if this actually helps but it's at least nice to make the conceptual separation.